README.md
# Program: socket
#### Original Source: apt-get source socket

## Description:
The socket 1.1 program can be trivally used to create a server.  This server can listen on either a TCP or a UNIX domain socket, and when a connection to this socket is made, a specific program can be executed.  Under the current implementation, a separate process is created (forked) when  a socket connection is made.

### Example:
$ socket -p "cat" -l -f -B localhost -s 7000

The above command line implements an 'echo' server.  The socket program creates a server-side (-s) socket on port 7000 bound (-B) to the localhost address. When the program accepts a network connection it forks (-f) a child proccess and then loops (-l) around to accept another connection.  The child process then executes a  program (-p), 'cat', in which its stdin and stdout is wired through the socket.

## Proposed Changes:
In this project, we seek to modify the existing socket program to create a master node that manages a set of workers. An initial set of workers are created (forked) at the start of the program.  As new requests are received, one of the workers fields the request.  Once complete, the work exits and the master preforks another worker.

Via this approach, there is always a set of workers ready to respond to a network request. As such, the turn-around time associated with each request is reduced, since the associated process-creation cost is not incurred.


### Example:
$ socket -p "cat" -l -f -W 4 -B localhost -s 7000

Notice in the above command line, we have added the -W option, with a value of 4.  The socket program will now maintain a pool of 4 workers.  The general operation of the program, the general flow of the master node is as follows:

* Create a socket: s=socket(AF_INET, SOCK_STREAM, 0)
* Bind the socket to the host address and port:  bind(s, ...)
* Listen for a network connection on the socket: listen(s, ...)
* Create a pool of workers: fork() 
* LOOP
  * Wait for a child process to complete: wait()
  * Replace the child process: fork()

Whereas the flow of each child node (i.e., worker) is as follows:

   * Accept the connection on the network socket: accept()
   * Rewire the file descriptors
   * Execute the defined program: execl("/bin/sh", "sh", "-c", "cat", NULL)
   * Exit the program: signal: exit()


## Planned Usage:
We are currently working on two other project that will immediately use the revised socket program.  The two projects are:

* [SCGI Daemon](https://github.com/csuntechlab/scgi-daemon)
* [FCGI Daemon](https://github.com/csuntechlab/fcgi-daemon)

Both these projects already use the socket program, within the scgi-launch and fcgi-launch scripts.  The use is als follows:

* socket -B ${ADDR} -s ${PORT} -b -f -q -l -p "${FCGI_WRAPPER} ${CGI_PROGRAM}"


## Additional Enhancements:
* Upon a successful accept(), have the child send a SIGCHILE to the master node for it to create a new worker process.
* Add the option to provide both a min and a max for the number of workers (Incremental Change)
  * Create the minimum number of workers at program startup
  * Add two additional workers for each completing worker, until the max is reached
  * Decrease the number of workers if no new requests are received over the lst U Units of time, unless the min is reached
* Create a ramp up strategy to address an expected spike of requests
  * Create a single worker at the program startup
  * Expand to the maxium number of works upon an accept() when the number of works is equal to the minimum
  * Decrease the number of workers over time



socket-1.1/CHANGES
-*-text-*-

User-visible changes in Socket version 1.1:

  - Ported to more Platforms (SGI Irix, IBM AIX, CDC EP/IX, ISC 3.0,
    System V R4, Sequent)

  - fixed getopt()-related bug in main() (opt declared as int now).

  - fixed SIGWINCH-related bug (socket no longer exits on SIGWINCH).

  - Background mode (-b option)

  - -f option to fork child 

  - fixed bug in server mode concerning the "Connection from..."
    message

  - slightly more adequate signal handling and messages

  - support of UNIX domain sockets


socket-1.1/io.c
/*

$Header: io.c[1.10] Sun Aug 30 19:21:18 1992 nickel@cs.tu-berlin.de proposed $
This file is part of socket(1).
Copyright (C) 1992 by Juergen Nickelsen <nickel@cs.tu-berlin.de>
Please read the file COPYRIGHT for further details.

*/

#define _BSD			/* AIX *loves* this */

#ifdef __linux__
#include <unistd.h>
#include <string.h>
#endif
#include <stdlib.h>
#include <sys/types.h>
#include <sys/time.h>
#ifdef ISC
#include <sys/bsdtypes.h>
#endif
#include <errno.h>
#include <stdio.h>
#include "globals.h"

/* read from from, write to to. select(2) has returned, so input
 * must be available. */
int do_read_write(from, to)
int from, to ;
{
    int size ;
    char input_buffer[BUFSIZ] ;
    
    if ((size = read(from, input_buffer, BUFSIZ)) == -1) {
	perror2("read") ;
	return -1 ;
    }
    if (size == 0) {		/* end-of-file condition */
	if (from == active_socket) {
	    /* if it was the socket, the connection is closed */
	    if (verboseflag) {
		fprintf(stderr, "connection closed by peer\n") ;
	    }
	    return -1 ;
	} else {
	    if (quitflag) {
		/* we close connection later */
		if (verboseflag) {
		    fprintf(stderr, "connection closed\n") ;
		}
		return -1 ;
	    } else if (verboseflag) {
		fprintf(stderr, "end of input on stdin\n") ;
	    }
	    readonlyflag = 1 ;
	    return 1 ;
	}
    }
    return do_write(input_buffer, size, to) ;

}

/* write the buffer; in successive pieces, if necessary. */
int do_write(buffer, size, to)
char *buffer ;
int size, to ;
{
    char buffer2[2 * BUFSIZ] ;	/* expanding lf's to crlf's can
				 * make the block twice as big at most */
    int written ;

    if (crlfflag) {
	if (to == active_socket) {
	    add_crs(buffer, buffer2, &size) ;
	} else {
	    strip_crs(buffer, buffer2, &size) ;
	}
    // debfix } else {
	// debfix bcopy(buffer, buffer2, size) ;
	buffer = buffer2 ;
    }
    while (size > 0) {
	// written = write(to, buffer2, size) ;
	written = write(to, buffer, size) ;
	if (written == -1) {
	    /* this should not happen */
	    perror2("write") ;
	    fprintf(stderr, "%s: error writing to %s\n",
		    progname,
		    to == active_socket ? "socket" : "stdout") ;
	    return -1 ;
	}
	size -= written ;
	buffer += written ;
    }
    return 1 ;
}

/* all IO to and from the socket is handled here. The main part is
 * a loop around select(2). */
void do_io()
{
    fd_set readfds ;
    int fdset_width ;
    int selret ;

    fdset_width = (IN > active_socket ? IN : active_socket) + 1 ;
    while (1) {			/* this loop is exited sideways */
	/* set up file descriptor set for select(2) */
	FD_ZERO(&readfds) ;
	if (!readonlyflag) {
	    FD_SET(IN, &readfds) ;
	}
	if (!writeonlyflag) {
	    FD_SET(active_socket, &readfds) ;
	}

	do {
	    /* wait until input is available */
	    selret = select(fdset_width, &readfds, NULL, NULL, NULL) ;
	    /* EINTR happens when the process is stopped */
	    if (selret < 0 && errno != EINTR) {
		perror2("select") ;
		exit(1) ;
	    }
	} while (selret <= 0) ;

	/* do the appropriate read and write */
	if (FD_ISSET(active_socket, &readfds)) {
	    if (do_read_write(active_socket, OUT) < 0) {
		break ;
	    }
	} else {
	    if (do_read_write(IN, active_socket) < 0) {
		break ;
	    }
	}
    }
}


socket-1.1/utils.c
/*

$Header: utils.c[1.14] Wed Sep  9 16:34:03 1992 nickel@cs.tu-berlin.de proposed $
This file is part of socket(1).
Copyright (C) 1992 by Juergen Nickelsen <nickel@cs.tu-berlin.de>
Please read the file COPYRIGHT for further details.

*/

#ifdef sgi
#define _BSD_SIGNALS
#define SIG_HANDLER_RET int
#else /* !sgi */
#define SIG_HANDLER_RET void
#endif

#include <stdio.h>
#ifdef __linux__
#include <unistd.h>
#endif
#include <stdlib.h>
#include <signal.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#ifdef ISC
#define WNOHANG 1
#else
#include <sys/resource.h>
#endif
#include "globals.h"


/* Signal handler, print message and exit */
SIG_HANDLER_RET exitsig(sig)
int sig ;
{
    if (sig != SIGUSR1) {
	fprintf(stderr, "\n%s occured, exiting\n", socket_siglist[sig]) ;
    }
    exit(-sig) ;
}

/* Give usage message */
void usage()
{
    static char ustring[] =
	"Usage: %s [-bclqrvw] [-B local ip] [-p prog] {{-s|host} port | [-s] /path}\n" ;

    fprintf(stderr, ustring, progname) ;
}

/* perror with progname */
void perror2(s)
char *s ;
{
    fprintf(stderr, "%s: ", progname) ;
    perror(s) ;
}

/* is s a number? */
int is_number(s)
char *s ;
{
    while (*s) {
	if (*s < '0' || *s > '9') {
	    return 0 ;
	}
	s++ ;
    }
    return 1 ;
}

/* set up signal handling. All except TSTP, CONT, CLD, and QUIT
 * are caught with exitsig(). */
void init_signals()
{
    int i ;
#ifdef BSD_SIG_SETMASK		/* only with BSD signals */
    static struct sigvec svec = { exitsig, ~0, 0 } ;
#endif

    initialize_siglist() ;	/* shamelessly stolen from BASH */
    
    for (i = 0; i < NSIG; i++) {
	switch (i) {
#ifdef SIGTSTP
	  case SIGTSTP:
	  case SIGTTOU:
	  case SIGTTIN:
	  case SIGSTOP:
	  case SIGCONT:
	    continue ;
#endif
#if !defined (SIGCHLD) && defined (SIGCLD)
#define SIGCHLD SIGCLD
#endif
#ifdef SIGCHLD
	  case SIGCHLD:
	    continue ;
#endif
#ifdef SIGWINCH
	  case SIGWINCH:	/* it is ridiculous to exit on WINCH */
	    continue ;
#endif
	  case SIGQUIT:		/* if the user wants a core dump, */
	    continue ;		/* they can have it. */
	  default:	    
#ifdef BSD_SIG_SETMASK
	    sigvec(i, &svec, NULL) ;
#else
	    signal(i, exitsig) ;
#endif
	}
    }
}

/* connect stdin with prog's stdout/stderr and stdout
 * with prog's stdin. */
void open_pipes(prog)
char *prog ;
{
    int from_cld[2] ;		/* from child process */
    int to_cld[2] ;		/* to child process */

    /* create pipes */
    if (pipe(from_cld) == -1) {
	perror2("pipe") ;
	exit(errno) ;
    }
    if (pipe(to_cld) == -1) {
	perror2("pipe") ;
	exit(errno) ;
    }

    /* for child process */
    switch (fork()) {
      case 0:			/* this is the child process */
	/* connect stdin to pipe */
	close(0) ;
	close(to_cld[1]) ;
	dup2(to_cld[0], 0) ;
	close(to_cld[0]) ;
	/* connect stdout to pipe */
	close(1) ;
	close(from_cld[0]) ;
	dup2(from_cld[1], 1) ;
	/* connect stderr to pipe */
	close(2) ;
	dup2(from_cld[1], 2) ;
	close(from_cld[1]) ;
	/* call program via sh */
	execl("/bin/sh", "sh", "-c", prog, NULL) ;
	perror2("exec /bin/sh") ;
	/* terminate parent silently */
	kill(getppid(), SIGUSR1) ;
	exit(255) ;
      case -1:
	perror2("fork") ;	/* fork failed */
	exit(errno) ;
      default:			/* parent process */
	/* connect stderr to pipe */
	close(0) ;
	close(from_cld[1]) ;
	dup2(from_cld[0], 0) ;
	close(from_cld[0]) ;
	/* connect stderr to pipe */
	close(1) ;
	close(to_cld[0]) ;
	dup2(to_cld[1], 1) ;
	close(to_cld[1]) ;
    }
}

/* remove zombie child processes */
void wait_for_children()
{
    int wret, status ;
#ifndef ISC
    struct rusage rusage ;
#endif

    /* Just do a wait, forget result */
#ifndef ISC
    while ((wret = wait3(&status, WNOHANG, &rusage)) > 0) ;
#else
    while ((wret = waitpid(-1, &status, WNOHANG)) > 0) ;
#endif
}

/* expand LF characters to CRLF and adjust *sizep */
void add_crs(from, to, sizep)
char *from, *to ;		/* *from is copied to *to */
int *sizep ;
{
    int countdown ;		/* counter */

    countdown = *sizep ;
    while (countdown) {
	if (*from == '\n') {
	    *to++ = '\r' ;
	    (*sizep)++ ;
	}
	*to++ = *from++ ;
	countdown-- ;
    }
}

/* strip CR characters from buffer and adjust *sizep */
void strip_crs(from, to, sizep)
char *from, *to ;		/* *from is copied to *to */
int *sizep ;
{

    int countdown ;		/* counter */

    countdown = *sizep ;
    while (countdown) {
	if (*from == '\r') {
	    from++ ;
	    (*sizep)-- ;
	} else {
	    *to++ = *from++ ;
	}
	countdown-- ;
    }
}

#define NULL_DEVICE "/dev/null"

/* put yourself in the background */
void background()
{
    int child_pid ;		/* PID of child process */
    int nulldev_fd ;		/* file descriptor for null device */

    child_pid = fork() ;
    switch (child_pid) {
      case -1:
	perror2("fork") ;
	exit(1) ;
      case 0:
#ifdef NOSETSID
	ioctl(0, TIOCNOTTY, 0) ;
#else
	setsid() ;
#endif
	chdir("/") ;
	if ((nulldev_fd = open(NULL_DEVICE, O_RDWR, 0)) != -1) {
	    int i ;

	    for (i = 0; i < 3; i++) {
		if (isatty(i)) {
		    dup2(nulldev_fd, i) ;
		}
	    }
	    close(nulldev_fd) ;
	}
	break ;
      default:
	exit(0) ;
    }
}


socket-1.1/globals.h
/*
  
$Header: globals.h[1.10] Sun Aug 30 21:13:41 1992 nickel@cs.tu-berlin.de proposed $
This file is part of socket(1).
Copyright (C) 1992 by Juergen Nickelsen <nickel@cs.tu-berlin.de>
Please read the file COPYRIGHT for further details.

*/

#include "patchlevel.h"
#include <errno.h>

/* globals for socket */

#define IN	0		/* standard input */
#define OUT	1		/* standard output */

#define LLEN	100		/* buffer size fo perror2() */

#ifdef __STDC__
#define A(args) args
#else
#define A(args) ()
#endif

int create_server_socket_inet A((long int local_ip, int port, int queue_length)) ;
int create_server_socket_unix A((char *pathname,int queue_length)) ;
int create_client_socket_inet A((long int local_ip, char **hostname, int port)) ;
int create_client_socket_unix A((char *pathname)) ;
int resolve_service A((char *name_or_number, char *protocol, char **name)) ;
void catchsig A((int sig)) ;
void usage A((void)) ;
int do_read_write A((int from, int to)) ;
int do_write A((char *buffer, int size, int to)) ;
void do_io A((void)) ;
char *so_release A((void)) ;
void open_pipes A((char *prog)) ;
void wait_for_children A((void)) ;
void perror2 A((char *s)) ;
int is_number A((char *s)) ;
void init_signals A((void)) ;
void add_crs A((char *from, char *to, int *sizep)) ;
void strip_crs A((char *from, char *to, int *sizep)) ;
void background A((void)) ;
void initialize_siglist A((void)) ;

//extern int errno ;

/* global variables */
extern int serverflag ;
extern int loopflag ;
extern int verboseflag ;
extern int readonlyflag ;
extern int writeonlyflag ;
extern int quitflag ;
extern int crlfflag ;
extern int active_socket ;
extern char *progname ;
extern char *socket_siglist[] ;


socket-1.1/siglist.c
/* siglist.c -- signal list for those machines that don't have one. */

/* Copyright (C) 1989 Free Software Foundation, Inc.

This file is part of GNU Bash, the Bourne Again SHell.

Modified by Juergen Nickelsen <nickel@cs.tu-berlin.de> for use with
Socket-1.1.

Bash is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 1, or (at your option) any later
version.

Bash is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License along
with Bash; see the file COPYING.  If not, write to the Free Software
Foundation, 675 Mass Ave, Cambridge, MA 02139, USA. */

#include <stdio.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <memory.h>

#if !defined (NSIG)
#  if defined (_NSIG)
#    define NSIG _NSIG
#  else
#    define NSIG 64
#  endif /* !_NSIG */
#endif /* !NSIG */

char *socket_siglist[NSIG];

/* extern *malloc (); */	/* see stdlib.h and/or malloc.h, by apal@szofi */

void initialize_siglist ()
{
  register int i;

  for (i = 0; i < NSIG; i++)
    socket_siglist[i] = (char *)0x0;

  socket_siglist[0] = "Bogus signal";

#if defined (SIGHUP)
  socket_siglist[SIGHUP] = "Hangup signal";
#endif

#if defined (SIGINT)
  socket_siglist[SIGINT] = "Interrupt";
#endif

#if defined (SIGQUIT)
  socket_siglist[SIGQUIT] = "Quit signal";
#endif

#if defined (SIGILL)
  socket_siglist[SIGILL] = "Illegal instruction";
#endif

#if defined (SIGTRAP)
  socket_siglist[SIGTRAP] = "BPT trace/trap";
#endif

#if defined (SIGIOT) && !defined (SIGABRT)
#define SIGABRT SIGIOT
#endif

#if defined (SIGABRT)
  socket_siglist[SIGABRT] = "ABORT instruction";
#endif

#if defined (SIGEMT)
  socket_siglist[SIGEMT] = "EMT instruction";
#endif

#if defined (SIGFPE)
  socket_siglist[SIGFPE] = "Floating point exception";
#endif

#if defined (SIGKILL)
  socket_siglist[SIGKILL] = "Kill signal";
#endif

#if defined (SIGBUS)
  socket_siglist[SIGBUS] = "Bus error";
#endif

#if defined (SIGSEGV)
  socket_siglist[SIGSEGV] = "Segmentation fault";
#endif

#if defined (SIGSYS)
  socket_siglist[SIGSYS] = "Bad system call";
#endif

#if defined (SIGPIPE)
  socket_siglist[SIGPIPE] = "Broken pipe condition";
#endif

#if defined (SIGALRM)
  socket_siglist[SIGALRM] = "Alarm clock signal";
#endif

#if defined (SIGTERM)
  socket_siglist[SIGTERM] = "Termination signal";
#endif

#if defined (SIGURG)
  socket_siglist[SIGURG] = "Urgent IO condition";
#endif

#if defined (SIGSTOP)
  socket_siglist[SIGSTOP] = "Stop signal";
#endif

#if defined (SIGTSTP)
  socket_siglist[SIGTSTP] = "Stopped";
#endif

#if defined (SIGCONT)
  socket_siglist[SIGCONT] = "Continue signal";
#endif

#if !defined (SIGCHLD) && defined (SIGCLD)
#define SIGCHLD SIGCLD
#endif

#if defined (SIGCHLD)
  socket_siglist[SIGCHLD] = "Child signal";
#endif

#if defined (SIGTTIN)
  socket_siglist[SIGTTIN] = "Stop (tty input) signal";
#endif

#if defined (SIGTTOU)
  socket_siglist[SIGTTOU] = "Stop (tty output) signal";
#endif

#if defined (SIGIO)
  socket_siglist[SIGIO] = "I/O ready signal";
#endif

#if defined (SIGXCPU)
  socket_siglist[SIGXCPU] = "CPU limit exceeded";
#endif

#if defined (SIGXFSZ)
  socket_siglist[SIGXFSZ] = "File limit exceeded";
#endif

#if defined (SIGVTALRM)
  socket_siglist[SIGVTALRM] = "Alarm (virtual)";
#endif

#if defined (SIGPROF)
  socket_siglist[SIGPROF] = "Alarm (profile)";
#endif

#if defined (SIGWINCH)
  socket_siglist[SIGWINCH] = "Window change";
#endif

#if defined (SIGLOST)
  socket_siglist[SIGLOST] = "Record lock signal";
#endif

#if defined (SIGUSR1)
  socket_siglist[SIGUSR1] = "User signal 1";
#endif

#if defined (SIGUSR2)
  socket_siglist[SIGUSR2] = "User signal 2";
#endif

#if defined (SIGMSG)
  socket_siglist[SIGMSG] = "HFT input data pending signal";
#endif 

#if defined (SIGPWR)
  socket_siglist[SIGPWR] = "power failure imminent signal";
#endif 

#if defined (SIGDANGER)
  socket_siglist[SIGDANGER] = "system crash imminent signal";
#endif 

#if defined (SIGMIGRATE)
  socket_siglist[SIGMIGRATE] = "Process migration";
#endif 

#if defined (SIGPRE)
  socket_siglist[SIGPRE] = "Programming error signal";
#endif 

#if defined (SIGGRANT)
  socket_siglist[SIGGRANT] = "HFT monitor mode granted signal";
#endif 

#if defined (SIGRETRACT)
  socket_siglist[SIGRETRACT] = "HFT monitor mode retracted signal";
#endif 

#if defined (SIGSOUND)
  socket_siglist[SIGSOUND] = "HFT sound sequence has completed signal";
#endif 

  for (i = 0; i < NSIG; i++)
    {
      if (!socket_siglist[i])
	{
	  socket_siglist[i] =
	    (char *) malloc (10 + strlen ("Unknown Signal #"));

	  sprintf (socket_siglist[i], "Unknown Signal #%d", i);
	}
    }
}


socket-1.1/socket.c
/*

$Header: socket.c[1.17] Wed Sep  9 16:23:14 1992 nickel@cs.tu-berlin.de proposed $
This file is part of socket(1).
Copyright (C) 1992 by Juergen Nickelsen <nickel@cs.tu-berlin.de>
Please read the file COPYRIGHT for further details.

*/

#ifdef __linux__
#include <unistd.h>
#endif 
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <sys/un.h>
#include <netdb.h>
#include <errno.h>
#include <stdio.h>
#ifdef SEQUENT
#include <strings.h>
#else
#include <string.h>
#endif
#include "globals.h"
#include <stdint.h> // debfix
#include <stdlib.h>

/* global variables */
int forkflag = 0 ;		/* server forks on connection */
int serverflag = 0 ;		/* create server socket */
int loopflag = 0 ;		/* loop server */
int verboseflag = 0 ;		/* give messages */
int readonlyflag = 0 ;		/* only read from socket */
int writeonlyflag = 0 ;		/* only write to socket */
int quitflag = 0 ;		/* quit connection on EOF */
int crlfflag = 0 ;		/* socket expects and delivers CRLF */
int backgflag = 0 ;		/* put yourself in background */
int active_socket ;		/* socket with connection */
char *progname ;		/* name of the game */
char *pipe_program = NULL ;	/* program to execute in two-way pipe */
int WaitForever = 0;            /* wait forever for socket on connection refused */
long int local_ip=0;           /* IP to bind at local */

void server A((long int local_ip,char *path, int port, char *service_name)) ;
void handle_server_connection A((void)) ;
void client A((long int local_ip,char *host, int port, char *service_name)) ;

int main(argc, argv)
int argc ;
char **argv ;
{
    char *cp ;			/* to point to '/' in argv[0] */
    int opt ;			/* option character */
    int error = 0 ;		/* usage error occurred */
    extern int optind ;		/* from getopt() */
/*    char *host ;  */		/* name of remote host */
    int port ;			/* port number for socket */
    char *service_name ;	/* name of service for port */
    char *path;			/* path to unix domain socket */

    /* print version ID if requested */
    if (argv[1] && !strcmp(argv[1], "-version")) {
	puts(so_release()) ;
	exit(0) ;
    }

    /* set up progname for later use */
    progname = argv[0] ;
    if ((cp = strrchr(progname, '/'))) progname = cp + 1 ;

    /* parse options */
    while ((opt = getopt(argc, argv, "B:bcflp:qrsvWw?")) != -1) {
	switch (opt) {
	  case 'f':
	    forkflag = 1 ;
	    break ;
	  case 'c':
	    crlfflag = 1 ;
	    break ;
          case 'B': 
            {
              struct hostent *dataip;
              char *cad;
              cad=argv[optind-1];
              dataip=gethostbyname(cad);
              if (dataip!=NULL) {
                local_ip=*((long int*)dataip->h_addr_list[0]);
              }
            }
            break;
	  case 'w':
	    writeonlyflag = 1 ;
	    break ;
	  case 'p':
	    pipe_program = argv[optind - 1] ;
	    break ;
	  case 'q':
	    quitflag = 1 ;
	    break ;
	  case 'r':
	    readonlyflag = 1 ;
	    break ;
	  case 's':
	    serverflag = 1 ;
	    break ;
	  case 'v':
	    verboseflag = 1 ;
	    break ;
	  case 'l':
	    loopflag = 1 ;
	    break ;
	  case 'b':
	    backgflag = 1 ;
	    break ;
	  case 'W':
	    WaitForever = 1 ;
	  default:
	    error++ ;
	}
    }
    if ( error || optind >= argc )
     {	usage();
	exit(15);
     }
    if ( strchr(argv[optind],'/') != NULL )
	path=argv[optind];
    else
	path=NULL;

    /* number of args ok? */
    if ( ( path == NULL && argc - optind + serverflag != 2 ) ||
    ( path != NULL && argc - optind != 1 ) )
     {	usage() ;
	exit(15) ;
     }

    /* check some option combinations */
#define senseless(s1, s2) \
    fprintf(stderr, "It does not make sense to set %s and %s.\n", (s1), (s2))

    if (writeonlyflag && readonlyflag) {
	senseless("-r", "-w") ;
	exit(15) ;
    }
    if (loopflag && !serverflag) {
	senseless("-l", "not -s") ;
	exit(15) ;
    }
    if (backgflag && !serverflag) {
 	senseless("-b", "not -s") ;
 	exit(15) ;
    }
    if (forkflag && !serverflag) {
	senseless("-f", "not -s") ;
    }
 
    /* set up signal handling */
    init_signals() ;

    /* get port number */
    if ( path==NULL )
     {    port = resolve_service(argv[optind + 1 - serverflag],
			   "tcp", &service_name) ;
	 if (port < 0) {
		fprintf(stderr, "%s: unknown service\n", progname) ;
		exit(5) ;
	    }
     }
    else
	port=-1;	/* no port for UNIX sockets */

    /* and go */
    if (serverflag) {
	if (backgflag) {
	    background() ;
	}
	server(local_ip,path,port,service_name) ;
    } else if ( path != NULL )
	client(local_ip,path, -1, service_name) ;
    else
    	client(local_ip,argv[optind], port, service_name) ;
    exit(0) ;
}


void server(local_ip,path,port, service_name)
long int local_ip ;
char *path;
int port ;
char *service_name ;
{
    int socket_handle, alen ;

    /* allocate server socket */
    if ( path != NULL )
	    socket_handle = create_server_socket_unix(path,1);
    else
	    socket_handle = create_server_socket_inet(local_ip,port, 1) ;
    if (socket_handle < 0) {
	perror2("server socket") ;
	exit(1) ;
    }
    if (verboseflag) {
	if ( path==NULL )
	 {	fprintf(stderr, "inet: listening on port %d", port) ;
		if (service_name) {
		    fprintf(stderr, " (%s)", service_name) ;
		}
		fprintf(stderr, "\n") ;
	 }
	else
		fprintf(stderr, "unix: bound to %s\n",path);
    }

    /* server loop */
    do {
	union 
	 {	struct sockaddr_in sa ;
		struct sockaddr_un su ;
	 } peer;
	    
	alen = sizeof(peer) ;

	/* accept a connection */
	if ((active_socket = accept(socket_handle,
			  (struct sockaddr *) &peer,
			  &alen)) == -1) {
	    perror2("accept") ;
	} else {
	    /* if verbose, get name of peer and give message */
	    if (verboseflag)
	     {
		if ( peer.sa.sin_family==AF_INET )
	 	 {	struct hostent *he ;
			// debfix long norder ;
			uint32_t norder;
			char dotted[20] ;

			he = gethostbyaddr((char *) &peer.sa.sin_addr.s_addr,
				   sizeof(peer.sa.sin_addr.s_addr), AF_INET) ;
			if (!he) {
			    norder = htonl(peer.sa.sin_addr.s_addr) ;
			    sprintf(dotted, "%ld.%ld.%ld.%ld",
				    (norder >> 24) & 0xff,
				    (norder >> 16) & 0xff,
				    (norder >>  8) & 0xff,
				    norder & 0xff) ;
			}
			fprintf(stderr, "inet: connection from %s\n",
				(he ? he->h_name : dotted)) ;
		 }
		else if ( peer.su.sun_family==AF_UNIX )
			fprintf(stderr,"unix: connection from %s\n",path);
	     }
	    if (forkflag) {
		switch (fork()) {
		  case 0:
		    handle_server_connection() ;
		    exit(0) ;
		  case -1:
		    perror2("fork") ;
		    break ;
		  default:
		    close(active_socket) ;
		    wait_for_children() ;
		}
	    } else {
		handle_server_connection() ;
	    }
	}
    } while (loopflag) ;

   close(socket_handle);
   if ( path != NULL )
	unlink(path);
	
}


void handle_server_connection()
{
    /* open pipes to program, if requested */
    if (pipe_program != NULL) {
	open_pipes(pipe_program) ;
    }
    /* enter IO loop */
    do_io() ;
    /* connection is closed now */
    close(active_socket) ;
    if (pipe_program) {
	/* remove zombies */
	wait_for_children() ;
    }
}


void client(local_ip,host, port, service_name)
long int local_ip;
char *host ;
int port ;
char *service_name ;
{
    int	is_unix_socket;

    /* get connection */
    if ( port>=0 )
     {    while (1) {
		active_socket = create_client_socket_inet(local_ip,&host, port) ;
		if (active_socket >= 0 || !WaitForever || errno != ECONNREFUSED) 
			break;
		sleep( 60 );
	    }
     }
    else
	   active_socket = create_client_socket_unix(host);	

    if (active_socket == -1) {
	perror2("client socket") ;
	exit(errno) ;
    } else if (active_socket == -2) {
	fprintf(stderr, "%s: unknown host %s\n", progname, host) ;
	exit(13) ;
    }
    if (verboseflag) {
	if ( port>=0 )
	 {	fprintf(stderr, "inet: connected to %s port %d", host, port) ;
		if (service_name) {
		    fprintf(stderr, " (%s)", service_name) ;
		}
		fprintf(stderr, "\n") ;
	 }
	else
		fprintf(stderr, "unix: connected to %s\n", host);
    }

    /* open pipes to program if requested */
    if (pipe_program != NULL) {
	open_pipes(pipe_program) ;
    }
    /* enter IO loop */
    do_io() ;
    /* connection is closed now */
    close(active_socket) ;
}

/*EOF*/
